### DISCUSSION TITLE
What is wrong with my complex BLE + Websocket code?

### DISCUSSION THREAD

**Discussion by kulcsarbence** [micropython/micropython#12381] [General]
Hello, I am using Raspberry Pico WH with a Motor.

I am trying to Start and Stop this motor using a websocket server, and I also want it to be able to be Started and Stopped with Bluetooth BLE as well. I am later gonna make an android app that will write to a specific BLE characteristic whether or not the motor should be started or stopped.

I have a few problems with my code:

1. The aioble.register_services(motor_service) runs just fine on boot, but then when I connect with Thonny and try to restart the program, it from then on always times out, I get ETIMEDOUT error every time except for boot.
2. I can connect to the BLE - advertised characteristic and write to it, it even starts the motor, but sometimes when I write a MOTOR_STOP, it does not respond. Hell, even the websocket and every functionality dies in that case, I cannot connect to the bluetooth anymore from that moment onward.
3. When on my websocket server I tell the Pico to start the motor, It works just well, then I tell it to STOP, it works as well. But here is the catch! Upon third or fourth time of doing this, the program stalls just again... Nothing works, not even the bluetooth connection.
4. If I send a Websocket START to my Pico, it starts the motor, then I try to connect to it with Bluetooth, sometimes the connection doesnt work, but even if I can connect to it with Bluetooth, it does not respond to me writing the Characteristic to STOP the motor. So the motor goes on indefinitely.
5. The biggest problem, If I do not use asyncio, Just run the websocket loop, or just run the advertise_and_wait SEPARATELY, then those just work fine... If I try to run the 3 async methods at once, It always produces some error...
6. Possible memory leak somewhere? My gc.mem_free() gets lower and lower as time goes on...

Here is my code:

https://pastecode.io/s/71x9ug2g

**Comment by Josverl** [comment]
With due respect, just sharing all our questions and code makes it hard to help.
I suggest to ask one q, get a solution for that, then move on to the next. Having your complete code is good, but it puts the effort on me to separate the relevant from the noise.

It helps us all in understanding and learning if you can build a minimal repro, showing the same problem

In fact, I often find that I can understand, and solve, my own issues in code that way.

**Reply by kulcsarbence** [reply]
Thanks, But the problem is that parts of my code work just well. Its their combination that makes all these errors, so grabbing parts of the code in this case I think is not a good approach for me :(

**Comment by andrewleech** [comment]
I suspect the problem may be `resp = websocket.recv()` as this to me looks like it'll be a blocking request? Any non-async blocking function will halt the async loop, preventing all concurrent async tasks from being serviced.

What is the `threadsafe` module you're using? I'm not sure if it's safe to run multiple `_thread.start_new_thread()` on rp2 as it only supports 1 extra thread (code running on processor core2). Maybe could try switching to https://github.com/Vovaman/micropython_async_websocket_client for the websocket?
